---
title: "WEHI data analysis challenge"
author: "Mei Du"
date: "08/01/2021"
output: html_document
---

### Set up and loading data
```{r packages, message=FALSE}
my_packages = c("limma", "Glimma", "edgeR", "biomaRt")
sapply(my_packages, require, character.only = TRUE)
options(stringsAsFactors = FALSE)
```

```{r data}
# Data packaging
data = read.table("Haemopedia-Human-RNASeq_raw.txt", header = TRUE) # read in count data
counts = as.matrix(data[,-1])
rownames(counts) = data[,1]

names = read.delim("Haemopedia-Human-RNASeq_samples.txt") # read sample data
ordered_names = names[match(colnames(counts), as.character(names[,1])),] # order sample data according to counts table column names 
group = factor(ordered_names[,2])
x = DGEList(counts = counts, group = group) # DGEList-object containing counts and gene annotations for all samples

geneid = data[,1]
mart = useMart(dataset="hsapiens_gene_ensembl",biomart='ensembl') # retrieve Ensembl gene annotations
genes = select(mart, keys = geneid, columns=c("ensembl_gene_id","uniprot_gn_symbol", "chromosome_name"), keytype="ensembl_gene_id")
ordered_genes = genes[match(geneid, genes$ensembl_gene_id),] # order annotation data according to counts table column names
ordered_genes$ensembl_gene_id = geneid
ordered_genes = ordered_genes[!duplicated(ordered_genes$ensembl_gene_id),] # remove any duplicate gene IDs
rownames(ordered_genes) = NULL
x$genes = ordered_genes # update gene information

x
```

### Data pre-processing
```{r process}
cpm = cpm(x)
lcpm = cpm(x, log=TRUE)
L = mean(x$samples$lib.size) * 1e-6
M = median(x$samples$lib.size) * 1e-6
c(L, M)
summary(lcpm)
```
### Question 1

#### What is the range of library sizes in the dataset?
```{r lib}
min(x$samples$lib.size)
max(x$samples$lib.size)
```
Library sizes range from 10,699,714 to 13,962,965.

### Question 2

#### For the smallest library, how many counts does a CPM value of 1 equate to? For the same library, what CPM cutoff would we use if we want to keep genes with 10 or more counts?
For the smallest library, a CPM value of 1 equates to having 10 counts, so the CPM cutoff would be 1. 0.94 could also work.

```{r cpm}
mil_reads = 10699714/(10^6)
cpm_thresh = 10 / mil_reads
cutoff = 0.94 * mil_reads
cutoff2 = 0.95 * mil_reads
cbind(mil_reads, cpm_thresh, cutoff, cutoff2)
```


### Question 3

#### Filter out lowly expressed genes using the filterByExpr function. Make a density plot of log-CPM values for pre-filtered and post-filtered data. How many genes were in the raw data and how many do we keep?

There were 60,504 genes and we keep 20,580.

```{r filter}
table(rowSums(x$counts==0)==nrow(x$samples)) 
keep = filterByExpr(x, group=group) # filter out lowly expressed genes
x_filtered=x[keep, keep.lib.sizes=FALSE]
dim(x_filtered)
```
##### Density plot
```{r density}
lcpm.cutoff <- log2(10/M + 2/L)
nsamples <- ncol(x)
library(scico)
col <- scico(nsamples, palette = "roma")

# pre-filtered raw data
plot(density(lcpm[,1]), col=col[1], lwd=1, ylim=c(0,3.4), las=2, main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
    den <- density(lcpm[,i])
    lines(den$x, den$y, col=col[i], lwd=1)
}
legend("topright",legend = ordered_names$sampleId, text.col=col, bty="n", cex=0.55, ncol=3)

# post-filtered data
lcpm_filtered <- cpm(x_filtered, log=TRUE)
plot(density(lcpm_filtered[,1]), col=col[1], lwd=1, ylim=c(0,3.4), las=2, main="", xlab="")
title(main="B. Filtered data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
    den <- density(lcpm_filtered[,i])
    lines(den$x, den$y, col=col[i], lwd=1)
}
legend("topright", legend = ordered_names$sampleId, text.col=col, bty="n",cex=0.55, ncol=3)
```


### Question 4
#### Comment on the scaling factors calculated in the normalisation step.

```{r norm}
x_filtered <- calcNormFactors(x_filtered, method = "TMM")
x_filtered$samples$norm.factors
```
Most of the scaling factors fall around 1, ranging from around 0.77 to 1.22, indicating a relatively small normalisation effect. Outliers included the 3 neutrophil samples, Neut.1 (0.4977324), Neut.2 (0.4829796) and Neut.3 (0.4524641).

### Question 5
#### Comment on the similarity/dissimilarity of samples based on an MDS plot. (Extra marks: Make an interactive MDS plot using the latest Glimma function, called glimmaMDS. Embed this plot into your report.)
```{r mds}
glMDSPlot(lcpm_filtered, labels=paste(rownames(x_filtered$samples)), groups = x_filtered$samples$group, launch = FALSE)
knitr::include_url("https://github.com/mei-du/data-challenge/blob/main/MDS-Plot.html")
```
The samples seem to cluster well within groups over dimension 1 and 2, and within cell lineages in particular. One contradicting instance is the 2 myDC samples that clustered with samples from the B cell lineage, separate from the other dendritic cell lineage samples. The biggest difference is observed between the B cell and T cell lineages over dimension 2. There is also a distinction between granulocytes and monocytes, and the lymphocytes over dimension 1, and further between granulocytes and monocytes over dimension 2. 

Another observation is the relatively low proportion of variance explained by dimensions 1 and 2, especially compared to the MDS plot in the paper. Perhaps the previous data collection and pre-processing work need to be investigated or there may be other factors affecting the analysis.


### Question 6
#### Comment on whether the mean-variance trend for this dataset is similar/dissimilar to the one from the RNAseq123 paper.
```{r design}
# model design
design = model.matrix(~0+group)
colnames(design) = gsub("group", "", colnames(design))
design

contr_mat = makeContrasts(
    BvsT =  (MemB+NveB) - (CD4T+CD8T),
    LymvsMonGran = (CD4T+CD8T+NK+MemB+NveB) - (Eo+Mono+MonoNonClassic+myDC+myDC123+Neut+pDC), 
    MonvsGran = (Mono+MonoNonClassic+myDC+myDC123+pDC) - (Eo+Neut),
    levels = colnames(design))
contr_mat
```

The mean-variance trend for this dataset, like the one from the paper, is decreasing, although in a more linear and less curved manner.

```{r mean-var}
# voom mean-variance 
v = voom(x_filtered, design, plot=TRUE)
title("voom: Mean-variance trend")
v

# fit linear model
vfit = lmFit(v, design)
vfit = contrasts.fit(vfit, contrasts=contr_mat)
efit = eBayes(vfit)
plotSA(efit)
title("Final model: Mean-variance trend")
```


### Question 7
#### How many differentially expressed genes are there using an adjusted p-value cutoff of 5%? What about at 10%?
```{r de}
summary(decideTests(efit))
```
For the comparison between expression levels in B and T cell genes, there are 10,596 differentially expressed genes. 16,878 DE genes are found between lymphocyte, and monocyte and granulocyte genes, and 17,337 DE genes are found between monocyte and granulocyte genes.

```{r p10}
summary(decideTests(efit, p.value = 0.1))
```

At a 10% adjusted p-value cutoff, there are 11,740 DE genes between B and T cell genes, 17,457 DE genes between lymphocyte, and monocyte and granulocyte gnees, and 17,848 DE genes between monocyte and granulocyte genes.

### Question 8
#### Using Glimma, make a mean-difference plot (using the glimmaMA function) that highlights/labels the top 5 most significant genes (by adjusted p-value). Save and embed this plot into your report.
```{r top}
# top DE genes
tfit = treat(vfit, lfc=1)
dt = decideTests(tfit)
summary(dt)

BvT = topTreat(tfit, coef=1, n=Inf)
LvMG = topTreat(tfit, coef=2, n=Inf)
MvG = topTreat(tfit, coef = 3, n = Inf)
head(BvT,5)
head(LvMG,5)
head(MvG,5)
```
```{r glimma}
glimmaMA(tfit, coef = 1, dge = x_filtered, groups = group)
glimmaMA(tfit, coef = 2, dge = x_filtered, groups = group)
glimmaMA(tfit, coef = 3, dge = x_filtered, groups = group)
```
